axt/axt.go

```diff
@@ -44,29 +45,39 @@ func Read(filename string) []Axt {
 	return answer
 }
 
-// ReadToChan takes a filename and a channel.  The function will read all remaining Axt records
+// ReadToChan takes an EasyReader and a channel.  The function will read all remaining Axt records
 // from the open file into the channel.  The function will then close the file and close the channel.
-func ReadToChan(filename string, data chan<- Axt) {
-	var file *fileio.EasyReader
+func ReadToChan(file *fileio.EasyReader, data chan<- Axt, wg *sync.WaitGroup) {
 	var curr Axt
 	var done bool
-	var err error
 
-	file = fileio.EasyOpen(filename)
 	for curr, done = ReadNext(file); !done; curr, done = ReadNext(file) {
 		data <- curr
 	}
-	err = file.Close()
-	exception.PanicOnErr(err)
-	close(data)
+
+	wg.Done()
 }
 
 // GoReadToChan takes a filename and returns a channel.  GoReadToChan
 // will launch a Go routine to read all the alignments from the file into the channel
 // and then close the channel.
-func GoReadToChan(filename string) <-chan Axt {
+func GoReadToChan(filename string) <-chan Axt { // filetypes w/ header would have another return here
+	var file *fileio.EasyReader
+	var err error
+	var wg sync.WaitGroup
+	file = fileio.EasyOpen(filename)
+
 	data := make(chan Axt, 1000)
-	go ReadToChan(filename, data)
+
+	wg.Add(1)
+	go func() { // spawn a goroutine that closes the file and channel when ReadToChan is finished
+		wg.Wait()
```

Comment:
I don't think this added complexity of having a go routine wait around for a signal to then close the file and close the channel buys us any advantages.  I think that when GoReadToChan finishes reading the file, it should just close the file and close the channel, instead of signalling to a waiting go routine to close the file and close the channel (we already know we are at the end of the file).

Reply:
This is a fair point. I considered letting ReadToChan close the file, but it seemed a bit messy to have ReadToChan silently close an already opened file. It seemed nicer to have open/close within the same function, but maybe that is not worth the complexity of another goroutine. As a last proposal I pushed another couple files that lay out how a header channel could work and we could have the best of both worlds. If that doesn't hold up to scrutiny, I will change it so ReadToChan closes the file and make it clear in the godoc that this happens.